paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreRead.java

```diff
@@ -267,17 +277,34 @@ private RecordReader<KeyValue> batchMergeRead(
         return keyProjectedFields == null ? reader : projectKey(reader, keyProjectedFields);
     }
 
-    private RecordReader<KeyValue> streamingConcat(
-            List<DataFileMeta> files, KeyValueFileReaderFactory readerFactory) throws IOException {
+    private RecordReader<KeyValue> noMergeRead(
+            BinaryRow partition,
+            int bucket,
+            List<DataFileMeta> files,
+            @Nullable List<DeletionFile> deletionFiles,
+            boolean onlyFilterKey)
+            throws IOException {
+        KeyValueFileReaderFactory readerFactory =
+                readerFactoryBuilder.build(
+                        partition, bucket, true, onlyFilterKey ? filtersForKeys : filtersForAll);
         List<ReaderSupplier<KeyValue>> suppliers = new ArrayList<>();
-        for (DataFileMeta file : files) {
+        for (int i = 0; i < files.size(); i++) {
+            DataFileMeta file = files.get(i);
+            DeletionFile deletionFile = deletionFiles == null ? null : deletionFiles.get(i);
             suppliers.add(
                     () -> {
                         // We need to check extraFiles to be compatible with Paimon 0.2.
                         // See comments on DataFileMeta#extraFiles.
                         String fileName = changelogFile(file).orElse(file.fileName());
-                        return readerFactory.createRecordReader(
-                                file.schemaId(), fileName, file.fileSize(), file.level());
+                        RecordReader<KeyValue> reader =
+                                readerFactory.createRecordReader(
+                                        file.schemaId(), fileName, file.fileSize(), file.level());
+                        if (deletionFile != null) {
+                            DeletionVector deletionVector =
+                                    DeletionVector.read(fileIO, deletionFile);
+                            reader = ApplyDeletionVectorReader.create(reader, deletionVector);
```

Comment:
Worry about conflicts with `readerFactoryBuilder.withDeletionVectorSupplier` in the future

Reply:
Only the ApplyDeletionVectorReader is the same part, and the Deletion is not conflicting and can be stacked.

In the future, we should avoid disturbing `ReaderFactoryBuilder` on the writer side and have independent deletion logic.